manager/dispatcher/dispatcher.go

```diff
@@ -1137,6 +1157,9 @@ func (d *Dispatcher) getRootCACert() []byte {
 // a special boolean field Disconnect which if true indicates that node should
 // reconnect to another Manager immediately.
 func (d *Dispatcher) Session(r *api.SessionRequest, stream api.Dispatcher_SessionServer) error {
+	d.shutdownWait.Add(1)
```

Comment:
@anshulpundir here and in a couple of places above, why isn't the check on `d.shutdown` required?

Reply:
Actually nevermind, after your most recent comment, this is not relevant.

Reply:
We're using the Dispatcher context to signal shutdown. 

Note the order: we always increment the waitgroup first, followed by isRunningLocked(). If an RPC finds the context not cancelled, it will have +1 the WaitGroup, which will make sure that the Stop() function waits for this rpc to finish.

Reply:
I would say that this is slightly clever, so make sure to document it somewhere that the ordering is what does the trick. It'll be hard to read later otherwise.